modules/numa-allocator/src/main/cpp/src/numa/numa_alloc.cpp

```diff
@@ -0,0 +1,196 @@
+        }
+
+        ~BitSetImpl() {
+            numa_bitmask_free(mask_);
+        }
+
+    private:
+        bitmask *mask_;
+        size_t size_;
+    };
+
+    BitSet::BitSet() {
+        p_impl_ = new BitSetImpl();
+    }
+
+    BitSet::BitSet(const BitSet &other) {
+        p_impl_ = new BitSetImpl(*other.p_impl_);
+    }
+
+    BitSet &BitSet::operator=(const BitSet &other) {
+        if (this != &other) {
+            BitSet tmp(other);
+            std::swap(this->p_impl_, tmp.p_impl_);
+        }
+        return *this;
+    }
+
+    bool BitSet::Get(size_t pos) const {
+        return p_impl_->GetBit(pos);
+    }
+
+    void BitSet::Set(size_t pos, bool value) {
+        p_impl_->SetBit(pos, value);
+    }
+
+    void BitSet::Set() {
+        p_impl_->SetAll(true);
+    }
+
+    void BitSet::Reset(size_t pos) {
+        p_impl_->SetBit(pos, false);
+    }
+
+    void BitSet::Reset() {
+        p_impl_->SetAll(false);
+    }
+
+    size_t BitSet::Size() const {
+        return p_impl_->Size();
+    }
+
+    bool BitSet::operator==(const BitSet &other) {
+        return this->p_impl_->Equals(*other.p_impl_);
+    }
+
+    bool BitSet::operator!=(const BitSet &other) {
+        return !(*this == other);
+    }
+
+    BitSet::~BitSet() {
+        delete p_impl_;
+    }
+
+    std::ostream &operator<<(std::ostream &os, const BitSet &set) {
+        os << '{';
+        for (size_t i = 0; i < set.Size(); ++i) {
+            os << set[i];
+            if (i < set.Size() - 1) {
+                os << ", ";
+            }
+        }
+        os << '}';
+        return os;
+    }
+
+    union region_size {
+        size_t size;
+        max_align_t a;
+    };
+
+    int NumaNodesCount() {
+        return numa_max_node() + 1;
+    }
+
+    template<typename Func, typename ...Args>
+    void *NumaAllocHelper(Func f, size_t size, Args ...args) {
+        auto ptr = static_cast<region_size *>(f(size + sizeof(region_size), args...));
+        if (ptr) {
+            ptr->size = size;
+            ptr++;
+        }
+        return ptr;
+    }
+
+    void *Alloc(size_t size) {
+        return NumaAllocHelper(numa_alloc, size);
+    }
+
+    void *Alloc(size_t size, int node) {
+        return NumaAllocHelper(numa_alloc_onnode, size, node);
+    }
+
+    void *AllocLocal(size_t size) {
+        return NumaAllocHelper(numa_alloc_local, size);
+    }
+
+    void *AllocInterleaved(size_t size) {
+        return NumaAllocHelper(numa_alloc_interleaved, size);
+    }
+
+    void *AllocInterleaved(size_t size, const BitSet &node_set) {
+        return NumaAllocHelper(numa_alloc_interleaved_subset, size, node_set.p_impl_->Raw());
+    }
+
+    size_t Size(void *buf) {
+        if (buf) {
+            auto *ptr = static_cast<region_size *>(buf);
+            ptr--;
+            return ptr->size;
+        }
+        return 0;
+    }
+
+    void Free(void *buf) {
+        if (buf) {
+            auto *ptr = static_cast<region_size *>(buf);
+            ptr--;
```

Comment:
I'd recommend writing a comment here and in `Size() ` about this cast and decrement, because ptr is not actually pointing to region_size before the decrement.